Source/inv.cpp

```diff
@@ -1109,6 +1110,37 @@ void CleanupItems(Item *item, int ii)
 	}
 }
 
+std::optional<Point> FindClosestValidItemPosition(Point startingPosition)
+{
+	if (CanPut(startingPosition)) {
+		return startingPosition;
+	}
+	if (dItem[startingPosition.x][startingPosition.y] == 0)
+		return {};
+
+	bool checkedTiles[MAXDUNX][MAXDUNY] = { false };
+	std::queue<Point> tilesToCheck;
+	tilesToCheck.push(startingPosition);
+	while (!tilesToCheck.empty()) {
+		Point currentPosition = tilesToCheck.front();
+		tilesToCheck.pop();
+		for (auto dir : PathDirs) {

```

Comment:
Had a play around with this and it turns out it's possible to drop items through stairs when diagonals are included -_-
https://user-images.githubusercontent.com/357684/147747604-624bd7d2-f4ca-4aa6-8612-665bb54ac8b9.mp4

I did a [rebase of this on current master](https://github.com/ephphatha/devilutionX/tree/qndel/fix_dropping_items) to test some changes, changing this line to only look at non-diagonal adjacent tiles avoids that issue.

```suggestion
		for (auto dir : { Direction::NorthEast, Direction::NorthWest, Direction::SouthEast, Direction::SouthWest }) {
```

Reply:
Well that is actually a general issue with the stairs, you can also see monsters on the other side of it etc.

Reply:
That's similar, in this case I'm mainly thinking of it from the way players path around solid tiles. Players can't walk diagonally between two tiles unless all four tiles that meet at that corner are non-solid (path_solid_pieces in path.cpp). This prevents players walking through stairs or between rocks.
![2021-12-30T233408_devilutionx](https://user-images.githubusercontent.com/357684/147752592-450c7c57-01f8-4b14-8aee-4f88bc522c72.png)

https://user-images.githubusercontent.com/357684/147752603-113cc65a-e0e0-4973-9d14-c571ce1dcd29.mp4
 Players are allowed to walk between two monsters/objects in the same formation though (visible at the end of that capture).

Having this check progress along north/south/east/west leads to item trails through gaps the player can't path through, only checking the four suggested directions avoids this issue and doesn't significantly impact dropping items near objects/corners.

Reply:
well for me it resulted in a significantly worse shape of the item pool